src/v/pandaproxy/schema_registry/json.cc

```diff
@@ -670,6 +670,94 @@ resolve_pointer(const json::Pointer& p, const json::Value& root) {
     return *value;
 }
 
+// builder for merged references. this will spilt out a json that is validated
+// by this schema
+// {
+//   "oneOf": [
+//     {
+//       "$comment": "the referenced schema, if the base schema did not contain
+//       siblings to $ref"
+//     },
+//     {
+//       "type": "object",
+//       "properties": {
+//         "allOf": {
+//           "type": "array",
+//           "minItems": 2,
+//           "items": true,
+//           "$comment": "items are 1) the original schema without the $ref and
+//           2) the referenced schema. if the referenced schema contains a $ref,
+//           further referenced schemas are flattened"
+//         }
+//       }
+//     }
+//   ]
+// }
+json_const_object
+merge_references(std::span<json::Value::ConstObject> references_objects) {
+    if (references_objects.empty()) {
+        throw std::logic_error("merged_references called on empty sequence");
+    }
+
+    const auto ref_key = json::Value{"$ref"};
+
+    auto empty_refless = [&](const json::Value::ConstObject& obj) {
+        // empty_refless if empty once $ref is removed
+        return obj.ObjectEmpty()
+               || (obj.MemberCount() == 1 && obj.FindMember(ref_key) != obj.MemberEnd());
+    };
+
+    // list of objects once we remove the $ref-only schemas
+    auto non_empty_references = references_objects
+                                | std::views::filter([&](const auto& obj) {
+                                      return !empty_refless(obj);
+                                  });
+
+    auto non_empty_size = std::ranges::distance(non_empty_references);
+
+    if (non_empty_size == 0) {
+        // a degenerate case: all the objects are empty, once removed the
+        // references.
+        return get_true_schema();
+    }
+
+    if (non_empty_size == 1) {
+        // only one object is not empty, return it directly
+        // this is the case for objects with a single $ref without siblings
+        return non_empty_references.front();
+    }
+
+    // at least 2 non-empty objects, build the allOf array
+
+    auto to_document_without_ref = [&](const json::Value::ConstObject& obj) {
+        // copy obj to a new document without the $ref key
+        auto doc = json::Document{rapidjson::kObjectType};
+        auto& alloc = doc.GetAllocator();
+        for (auto& [k, v] : obj) {
+            if (k != ref_key) {
+                doc.AddMember(
+                  json::Value(k, alloc), json::Value(v, alloc), alloc);
```

Comment:
note: I guess we don't need to copy any strings (`json::Value` 3rd param) since everything is kept alive in another doc somewhere.

Reply:
i tried but that's not how the rapidjson api works, the Document has to own <key,value>

Reply:
> i tried but that's not how the rapidjson api works, the Document has to own <key,value>

I thought the default was to not copy?

Reply:
https://github.com/Tencent/rapidjson/blob/815e6e7e7e14be44a6c15d9aefed232ff064cad0/include/rapidjson/document.h#L1381C9-L1387C36

basically it requires lvalues `Value::AddMember(Value&, Value&, Allocator&)` but the implementation will take ownership and set the inputs as null.
i haven't found another API to do it

Reply:
> https://github.com/Tencent/rapidjson/blob/815e6e7e7e14be44a6c15d9aefed232ff064cad0/include/rapidjson/document.h#L1381C9-L1387C36
> 
> basically it requires lvalues `Value::AddMember(Value&, Value&, Allocator&)` but the implementation will take ownership and set the inputs as null. i haven't found another API to do it

I was thinking of this: https://github.com/Tencent/rapidjson/blob/815e6e7e7e14be44a6c15d9aefed232ff064cad0/include/rapidjson/document.h#L742

Reply:
ah, you mean `bool copyConstStrings = false)`
do we dare? seems doable

Reply:
> ah, you mean `bool copyConstStrings = false)` do we dare? seems doable

Nothing to be done, I think.

Reply:
ok i added it

Reply:
> ok i added it

OK, but it was still the default! Explicit is fine.

Reply:
🤦 i wrote `false` in the code instead of `true`

ok i'll change it in the next PR

Reply:
> 🤦 i wrote `false` in the code instead of `true`
> 
> ok i'll change it in the next PR

I'm not asking you to change anything.  We're already NOT copying, that's the default. That seems more efficient. The first comment was ensuring that it's safe to take by reference, since the other one is still alive. You added comments to that effect.